### PR TITLE
Feature/251 quick indexing pipeline

### PR DESCRIPTION
**Pull request by K-Schubert**
Added scraping/indexing of PDFs from https://ahv-iv.ch memento section.

Added scraping/indexing of HTML webpages from https://www.eak.admin.ch and https://www.zas.admin.ch sitemaps.

NOTE: this PR is based on the refactored branch feature/236-integrate-os-llm. The utils dir has been renamed to components, and interfaces for embedding/llm models as well as factories have been added.

### CODE REVIEW

File: src/copilot/app/autocomplete/matching.py [hunk 1]
```diff
@@ -60,14 +59,24 @@ class FuzzyMatch(Matching):
     A class that implements fuzzy matching through levenshtein, return results with the lowest distance first. The
     levenshtein distance is defined as the number of changes required to get to the target string.
     """
-    def __init__(self, threshold: int = 100):
-        self.match_type = "exact_match"
+    def __init__(self, threshold: int = 50, fuzzy_limit: int = 0):
+        self.match_type = "fuzzy_match"
         self.threshold = threshold
+        self.fuzzy_limit = fuzzy_limit
```
Comment (Shi-Ho) [inline src/copilot/app/autocomplete/matching.py:65]: self.limit is setup in the parent class constructor

File: src/copilot/app/components/embedding/base.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        Returns
+        -------
+        list of list of float
+            The embeddings for the texts.
+        """
+        return [self.truncate_text(text) for text in texts]
+
+    @abstractmethod
+    async def aembed_query(self, text: str) -> List[float]:
```
Comment (Shi-Ho) [inline src/copilot/app/components/embedding/base.py:109]: typo

File: src/copilot/app/components/indexing/base.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
+"""**Indexing** interface."""
+import logging
+
+from abc import ABC, abstractmethod
+from typing import List, Any
+import aiohttp
+
+from haystack.dataclasses import Document
```
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/base.py:8]: Not a good practice to use an implementation specific library in the abstract class. Can we avoid using Document dataclass?

File: src/copilot/app/components/indexing/base.py
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/base.py:1]: I would move the whole app/components/indexing/ directory to app/indexing. Also we should have a clearer vision of what is the use of having different implementation of those abstract classes

File: src/copilot/app/components/indexing/factory.py
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/factory.py:1]: empty file

File: src/copilot/app/components/indexing/parsing.py
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/parsing.py:1]: why is this not in the parser class

File: src/copilot/app/components/indexing/implementations/haystack.py [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+    A class used to scrape URLs and extract content from them.
+
+    Attributes
+    ----------
+    fetcher : LinkContentFetcher
+        An instance of LinkContentFetcher to fetch the content of URLs.
+
+    Methods
+    -------
+    __init__()
```
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/implementations/haystack.py:35]: not necessary to list the constructor

File: src/copilot/app/components/indexing/implementations/haystack.py [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+
+    cleaner : DocumentCleaner
+        An instance of DocumentCleaner to clean documents.
+
+    splitter : DocumentSplitter
+        An instance of DocumentSplitter to split documents into chunks.
+
+    Methods
+    -------
+    __init__()
```
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/implementations/haystack.py:167]: not necessary

File: src/copilot/app/components/indexing/base.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        Returns
+        -------
+        list of Document
+            The converted documents.
+        """
+
+    @abstractmethod
+    def clean_documents(self, documents: List[Document]) -> List[Document]:
+        """
+        Abstract method to clean documents.
```
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/base.py:193]: + remove empty documents

File: src/copilot/app/components/indexing/implementations/haystack.py [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+        documents : List[Document]
+            The documents to split into chunks.
+
+        Returns
+        -------
+        List[Document]
+            A list of documents split into chunks.
+        """
+        # Remove empty documents
+        documents = super().clean_documents(documents)
```
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/implementations/haystack.py:325]: why don't you call clean_document from the same class

File: src/copilot/app/components/indexing/implementations/haystack.py [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+        # TO DO: refactor embedding logic to embed from documents (add from_documents method)
+        # Embed documents
+        str_chunks = [doc.content for doc in chunks["documents"]]
+        document_embeddings = self.embedding_client.embed_documents(str_chunks)
+        document_embeddings = [e.embedding for e in document_embeddings]
+        urls = [doc.meta["url"] for doc in chunks["documents"]]
+
+        # Upsert documents into VectorDB
+        for embedding, doc, url in zip(document_embeddings, str_chunks, urls):
+            date = datetime.now()
```
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/implementations/haystack.py:440]: datetime is done within the database already

File: src/copilot/app/components/indexing/implementations/haystack.py [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+    Methods
+    -------
+    __init__()
+        Initializes the HaystackIndexer with HaystackScraper, HaystackParser, and EmbeddingClient instances.
+    index_html_from_sitemap(sitemap_url: str, language: str = "de") -> dict
+        Indexes HTML content from the given sitemap URL into the VectorDB.
+    index_pdfs_from_sitemap(sitemap_url: str, language: str = "de") -> dict
+        Indexes PDF content from the given sitemap URL into the VectorDB.
+    """
+    def __init__(self):
```
Comment (Shi-Ho) [inline src/copilot/app/components/indexing/implementations/haystack.py:352]:
```
def __init__(self, scraper: BaseScraper, parser: BaseParser):
        """
        Initializes the HaystackIndexer with HaystackScraper, HaystackParser, and Embedding instances.
        """
        self.scraper = scraper
        self.parser = parser
        self.embedding_client = EmbeddingFactory.get_embedding_client(rag_config["embedding"]["model"])
```

File: src/copilot/app/components/indexing/implementations/haystack.py (hunk 7 above)
Comment (Shi-Ho) [reply src/copilot/app/components/indexing/implementations/haystack.py:325]: duplicate code

File: src/copilot/app/components/embedding/base.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+            The text to embed.
+
+        Returns
+        -------
+        list of float
+            The embedding for the text.
+        """
+
+    @abstractmethod
+    async def aembed_documents(self, texts: List[str]) -> List[List[float]]:
```
[the remark this answers is not in the document]
Comment (K-Schubert) [reply src/copilot/app/components/embedding/base.py:92]: Supposed to be async embed (naming convention generally used).

File: src/copilot/app/components/embedding/base.py (hunk 2 above)
Comment (K-Schubert) [reply src/copilot/app/components/embedding/base.py:109]: Same naming convention.

File: src/copilot/app/components/indexing/implementations/haystack.py (hunk 8 above)
Comment (K-Schubert) [reply src/copilot/app/components/indexing/implementations/haystack.py:440]: Removed date args for "created_at", "modified_at". Also changed the indexing/queries.py insert_rag() function to not intake these parameters:

```
async def insert_rag(embedding: str,
                     text: str,
                     url: str):
    conn = await get_db_connection()

    try:
        await conn.execute("""
            INSERT
            INTO embeddings (embedding, text, url)
            VALUES ($1, $2, $3)
        """, embedding, text, url)

    finally:
        await conn.close()
```

File: src/copilot/app/components/indexing/base.py [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+from typing import List, Any
+import aiohttp
+
+from haystack.dataclasses import Document
+
+# Setup logging
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+logger = logging.getLogger(__name__)
+
+class BaseScraper(ABC):
```
[the remark this answers is not in the document]
Comment (K-Schubert) [reply src/copilot/app/components/indexing/base.py]: Defined a "scrap_urls" and "fetch" method for the moment. The class will probably only be used to scrap from a URL.

[the remark this answers is not in the document]
Comment (K-Schubert) [reply src/copilot/app/components/indexing/base.py]: We always retrieve content as bytes. Then the parser should have the HTML/PDF/XML parsing functionality.

File: src/copilot/app/components/indexing/parsing.py
Comment (K-Schubert) [reply src/copilot/app/components/indexing/parsing.py:1]: changed to parser class

**Comment by K-Schubert** [comment]
@Shi-Ho Please check latest commits.

**Pull request merged**